src/NumFieldOrd/NfOrd/NfOrd.jl

```diff
@@ -999,114 +992,105 @@ The equation order of the number field.
+    # new_bas := [e, e*a_2, ..., e*a_k] and we append this to bas and set
+    # start := k + 1. In the next iteration, we then have
+    # new_bas := [e^2, e^2*a_2, ..., e^2*a_k] (assuming that there was no
+    # reduction of the basis in between).
+    for i in 1:df
+      new_bas = elem_type(K)[]
+      for j in start:length(bas)
+        t = e*bas[j]
+        in_span_of_B(t) && continue
+        push!(new_bas, t)
       end
+      isempty(new_bas) && break
+      start = length(bas) + 1
+      append!(bas, new_bas)
+
       if length(bas) >= n
+        # HNF reduce the basis we have so far, if B is already of full rank,
+        # we can do this with the modular algorithm
         B = basis_matrix(bas, FakeFmpqMat)
-        if extending
-          # We are extending extended_order, which has basis matrix M/d
-          # Thus we know that B.den/d * M \subseteq <B.num>
-          # So we can take B.den/d * largest_elementary_divisor(M) as the modulus
-          B = hnf_modular_eldiv(B, B.den, shape = :lowerleft)
+        if full_rank
+          # We have M/d \subseteq B, where M/d is a former incarnation of B.
+          # So we can use B.den*det(M) as a modulus.
+          hnf_modular_eldiv!(B, denominator(B, copy = false)*m, shape = :lowerleft)
+          B = sub(B, nrows(B) - n + 1:nrows(B), 1:n)
         else
           hnf!(B)
+          k = findfirst(k -> !is_zero_row(B, k), nrows(B) - n + 1:nrows(B))
+          B = sub(B, nrows(B) - n + k:nrows(B), 1:n)
+          if nrows(B) == n
+            full_rank = true
+            m = det(numerator(B, copy = false))
+          end
         end
-        rk = nrows(B) - n + 1
-        while is_zero_row(B, rk)
-          rk += 1
-        end
-        B = sub(B, rk:nrows(B), 1:n)
-        phase = 2
-        bas = elem_type(K)[ elem_from_mat_row(K, B.num, i, B.den) for i = 1:nrows(B) ]
+        bas = elem_type(K)[ elem_from_mat_row(K, numerator(B, copy = false), i, denominator(B, copy = false)) for i = 1:nrows(B) ]
+        start = 1
         if check
           @assert isone(bas[1])
         end
       end
     end
   end
-
-  if length(bas) > n # == n can only happen here after an hnf was computed
-                     # above. Don't quite see how > n can happen here either
-    B = basis_matrix(bas, FakeFmpqMat)
-    hnf!(B)
-    rk = nrows(B) - n + 1
-    if is_zero_row(B.num, rk)
-      error("data does not define an order: dimension to small")
-    end
-    B = sub(B, rk:nrows(B), 1:n)
-    bas = elem_type(K)[ elem_from_mat_row(K, B.num, i, B.den) for i = 1:nrows(B) ]
+  if length(bas) < n
+    error("The elements do not define an order: rank too small")
   end
-
-  if !isdefined(B, :num)
-    error("data does not define an order: dimension to small")
-  end
-
-  # Make an explicit check
-  @hassert :NfOrd 1 defines_order(K, B)[1]
-  return Order(K, B, cached = cached, check = check)
+  return Order(K, B, cached = cached, check = check)::order_type(K)
```

Comment:
Doesn't `check = false` or `check = (check && extends !== nothing)` suffice, depending on whether or not we're supposed to check that `extends` is actually an order?

You've already checked that the generators are integral.

The `Order` constructor with `check = true` reduces all products of pairs of basis elements modulo the (Hermite reduced) basis. For example if there is only one generator, this seems like an unnecessary amount of work.

Reply:
This depends on the general philosophy, I think. My understanding so far is that we by default rather check correctness one time too often, but let the user turn off the checks, if they know what they are doing.